Main.hs

```diff
@@ -71,67 +70,51 @@ instance MonadState EvalState EvalM where
-                                     , _recentMsgs = M.empty
-                                     , _channel = c
-                                     }
-          forkIO $ runReaderT (runEvalM backendLoop) env
-          runReaderT (runEvalM mainLoop) env
-  where mainLoop = forever $ catch (do bot <- liftIO $ loginRestGateway =<< getAuth
-                                       finally (do use botHandle >>= \h -> liftIO $ putMVar h bot
+          env <- newMVar $ EvalState { _botHandle = bh
+                         , _recentMsgs = M.empty
+                         , _channel = c
+                         }
+          tid <- forkIO $ runReaderT (runEvalM backendLoop) env
+          finally (runReaderT (runEvalM frontLoop) env)
+                  (killThread tid)
+
+frontLoop :: EvalM ()
+frontLoop = forever $ catch (do bot <- liftIO $ loginRestGateway =<< getAuth
+                                void $ finally (do use botHandle >>= \h -> liftIO $ putMVar h bot
                                                    logM "Connected"
                                                    eventLoop
                                                    use botHandle >>= \h -> liftIO $ takeMVar h)
                                                (do logM "Disconnected"
-                                                   liftIO $ stopDiscord bot)
-                                       pure ())
-                                   (\e -> do logM "Exception in mainLoop:"
-                                             logShowM (e :: SomeException)
-                                             liftIO $ threadDelay $ 10 * 1000000)
-        eventLoop = do bot <- use botHandle >>= liftIO . readMVar
-                       everr <- liftIO $ nextEvent bot
-                       case everr of
-                         Right event -> do catch (handleEvent event)
-                                                 (\e -> do logM "Exception in eventLoop:"
-                                                           logShowM (e :: SomeException))
-                                           eventLoop
-                         Left err -> do logM "Exception in nextEvent:"
-                                        logShowM err
+                                                   liftIO $ stopDiscord bot))
+                            (\e -> do logM "Exception in mainLoop:"
+                                      logM (show (e :: SomeException))
+                                      liftIO $ threadDelay $ 10 * 10^6)
+
+eventLoop :: EvalM ()
+eventLoop = do bot <- use botHandle >>= liftIO . readMVar
+               everr <- liftIO $ nextEvent bot
+               case everr of
+                 Right event -> do handleEvent event
+                                   eventLoop
+                 Left err -> do logM "Exception in nextEvent:"
+                                logM (show err)
+                                liftIO $ threadDelay $ 10 * 10^6
```

Comment:
If an exception occurs in the event loop we want to reconnect immediately. We only want to wait in the other spot because that spot is hit when there's an exception during *connection*, and we don't want to DoS the discord servers.

Reply:
I'll remove the `threadDelay` from the pull request